### ISSUE TITLE
Train generator every n_critic step

### ISSUE THREAD

**Issue by lucmos** [nocotan/pytorch-lightning-gans#7]
Hey, 
I'm wondering about this condition. If I understood correctly, the goal is to train the generator every `n_critic` steps.

https://github.com/nocotan/pytorch-lightning-gans/blob/ca38215807971d6cbce97cde47f669f3c107b611/models/wgan.py#L98-L109

The if condition at line 109 does not do that. If I am not mistaken:

- `optimizer_idx` can be either `1` or `0`. The opt `0` is the generator optimizer, the opt `1` is the discriminator
- At every step, `training_step` is called for each optimizer, with the correct `optimizer_idx`

Thus the condition `optimizer_idx % n_critic == 0` is equivalent to `if optimizer_idx == 0`. It trains the generator using the correct optimzer, but it does so at every step.

**Comment by nocotan** [comment]
@lucmos Thanks for pointing that out!

I agree with you and I think this is a bug.
The correct code may be as follows:
```python
def training_step(self, batch, batch_idx, optimizer_idx): 
     imgs, _ = batch 
  
     # sample noise 
     z = torch.randn(imgs.shape[0], self.latent_dim) 
     z = z.type_as(imgs) 
  
     n_critic = 5 
     clip_value = 0.01 
  
     # train generator 
     if optimizer_idx == 0 and batch_idx % n_critic ==0:
         ...
```

I did not notice this.
Thank you so much!

Would you be willing to make a pull request to fix this bug?
If not, I will fix this myself as soon as possible.

**Comment by lucmos** [comment]
@nocotan thanks for the reply! 

I agree with your fix, remember the `else` branch to return `None` (in order to skip the training for that optimizer in that step).  
Otherwise, sometimes it will train the generator with the loss computed for the discriminator (i.e. there will be a mismatch between losses and optimizers).

Something like this may be cleaner:

```python
def training_step(self, batch, batch_idx, optimizer_idx): 
    # train generator 
    if optimizer_idx == 0 and self.global_step % n_critic ==0:
        ...
    # train discriminator
    elif optimizer_idx == 1:
        ...
    else:
        return None
```

I'm sorry but atm I can't make a pull request :(

**Comment by nocotan** [comment]
@lucmos Thank you for your suggestion!

I think your fix is so beautiful.


Okay, I'll fix it myself.
Thank you for your advice!
